tests/stdlib/tasynchttpserver_transferencoding.nim

```diff
@@ -1,42 +1,56 @@
+discard """
+  matrix: "--gc:arc --threads:on; --gc:arc --threads:on -d:danger; --threads:on"
+"""
+
 import httpclient, asynchttpserver, asyncdispatch, asyncfutures
 import net
 
 import std/asyncnet
 import std/nativesockets
+import std/threadpool
 
 const postBegin = """
 POST / HTTP/1.1
 Transfer-Encoding:chunked
 
 """
 
-template genTest(input, expected) =
-  var sanity = false
-  proc handler(request: Request) {.async.} =
+template genTest(input, expected: string) =
+  proc handler(request: Request, future: Future[bool]): Future[void] {.async, gcsafe.} =
       doAssert(request.body == expected)
       doAssert(request.headers.hasKey("Transfer-Encoding"))
       doAssert(not request.headers.hasKey("Content-Length"))
-      sanity = true
+      future.complete(true)
       await request.respond(Http200, "Good")
 
-  proc runSleepLoop(server: AsyncHttpServer) {.async.} = 
+  proc runSleepLoop(server: AsyncHttpServer, future: Future[bool]) {.async, gcsafe.} = 
+    proc wrapper(request: Request): Future[void] {.gcsafe, closure.} =
+      # Capture the future
+      handler(request, future)
+    await server.acceptRequest(wrapper)
+
+  proc sendData(data: string, port: Port): void =
+    var socket = newSocket()
+    defer: socket.close()
+
+    socket.connect("127.0.0.1", port)
+    socket.send(data)
+
+  proc runTest(): Future[bool] {.async.} =
+    var handlerFuture = newFuture[bool]("runTest")
+    let data = postBegin & input
+    let server = newAsyncHttpServer()
     server.listen(Port(0))
-    proc wrapper() = 
-      waitFor server.acceptRequest(handler)
-    asyncdispatch.callSoon wrapper
 
-  let server = newAsyncHttpServer()
-  waitFor runSleepLoop(server)
-  let data = postBegin & input
-  var socket = newSocket()
-  socket.connect("127.0.0.1", server.getPort)
-  socket.send(data)
-  waitFor sleepAsync(10)
-  socket.close()
-  server.close()
+    spawn sendData(data, server.getPort)
```

Comment:
This can be async too, you shouldn't need to use `spawn`.

Reply:
Fixed.

Reply:
B1 if you want the socket connection and send to happen in the background.

B2 if you want to pause the current async proc waiting on the socket connection and send to happen.

Reply:
A1/A2: both are not ideal:

```nim
  proc sendData(data: string, port: Port) {.async.} =
    var socket = newAsyncSocket()
    defer: socket.close()
    await socket.connect("127.0.0.1", port)
    await socket.send(data)
```